In [115]:
import numpy as np
import pandas as pd
import plotly.express as px
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, \
    mean_absolute_percentage_error
from sklearn.model_selection import train_test_split


In [79]:
n_samples = 100
weight = np.random.normal(loc=50, scale=10, size=n_samples)  # Peso dos ratos em gramas
size = 30 + 0.5 * weight + np.random.normal(loc=0, scale=5, size=n_samples)
month_born = np.random.randint(1, 13, size=n_samples)
rats = {
    'weight': weight,
    'size': size,
    'month_born': month_born
}
dataframe = pd.DataFrame(rats)
dataframe

weight       size  month_born
0   60.960544  54.320974           5
1   67.605696  69.668205           2
2   40.750719  45.184430           9
3   58.727479  63.722893           6
4   55.562310  61.525420           4
..        ...        ...         ...
95  38.655585  55.191394           2
96  58.939880  63.824424          12
97  46.162430  54.751778           6
98  67.342723  67.196215           8
99  42.257558  44.342544           5

[100 rows x 3 columns]

In [80]:
fig = px.scatter(dataframe, x='size', y='weight', trendline='ols')
fig.show()

In [81]:
fig = px.scatter(dataframe, x='month_born', y='weight', trendline='ols')
fig.show()

## MSE

In [110]:
X, y = dataframe[['size']], dataframe['weight']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_train

size
11  56.268104
47  53.995523
85  44.154838
28  53.120752
93  58.164770
..        ...
60  57.699852
71  53.668665
14  62.522557
92  49.207768
51  62.879007

[70 rows x 1 columns]

In [111]:
y_train

11    53.104169
47    33.721066
85    51.413835
28    44.061383
93    47.481200
        ...    
60    45.588480
71    70.658471
14    45.121083
92    34.951816
51    48.052236
Name: weight, Length: 70, dtype: float64

In [112]:
lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression()

In [113]:
y_pred = lr.predict(X_test)

In [116]:
print(f'MSE using size to predict :{mean_squared_error(y_test, y_pred)}')
print(f'RMSE using size to predict :{root_mean_squared_error(y_test, y_pred)}')
print(f'MAE using size to predict :{mean_absolute_error(y_test, y_pred)}')
print(f'MAPE using size to predict :{mean_absolute_percentage_error(y_test, y_pred)}')

MSE using size to predict :32.092064606132304
RMSE using size to predict :5.664985843418526
MAE using size to predict :4.558531024817282
MAPE using size to predict :0.09270083592592529


In [118]:
X, y = dataframe[['month_born']], dataframe['weight']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_train

month_born
11           4
47           3
85          12
28          10
93           6
..         ...
60           9
71           4
14           9
92           1
51          12

[70 rows x 1 columns]

In [119]:
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)

In [120]:
print(f'MSE using month to born to predict :{mean_squared_error(y_test, y_pred)}')
print(f'RMSE using month to born to predict :{root_mean_squared_error(y_test, y_pred)}')
print(f'MAE using month to born to predict :{mean_absolute_error(y_test, y_pred)}')
print(f'MAPE using month to born to predict :{mean_absolute_percentage_error(y_test, y_pred)}')

MSE using month to born to predict :76.74583103091776
RMSE using month to born to predict :8.76046979510333
MAE using month to born to predict :6.929419052240534
MAPE using month to born to predict :0.16089084049913616


In [122]:
X, y = dataframe[['month_born', 'size']], dataframe['weight']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_train

month_born       size
11           4  56.268104
47           3  53.995523
85          12  44.154838
28          10  53.120752
93           6  58.164770
..         ...        ...
60           9  57.699852
71           4  53.668665
14           9  62.522557
92           1  49.207768
51          12  62.879007

[70 rows x 2 columns]

In [123]:
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)

In [124]:
print(f'MSE using month to born and size to predict :{mean_squared_error(y_test, y_pred)}')
print(f'RMSE using month to born and size to predict :{root_mean_squared_error(y_test, y_pred)}')
print(f'MAE using month to born and size to predict :{mean_absolute_error(y_test, y_pred)}')
print(f'MAPE using month to born and size to predict :{mean_absolute_percentage_error(y_test, y_pred)}')

MSE using month to born and size to predict :32.559334599604945
RMSE using month to born and size to predict :5.70607874109751
MAE using month to born and size to predict :4.61088154258837
MAPE using month to born and size to predict :0.09374353731896855


}MSE using size to predict :32.092064606132304 <br>
MSE using month to born to predict :76.74583103091776 <br>
MSE using month to born and size to predict :32.559334599604945 <br>
---------------------------------------------------------------------------
RMSE using size to predict :5.664985843418526<br>
RMSE using month to born to predict :8.76046979510333<br>
RMSE using month to born and size to predict :5.70607874109751
---------------------------------------------------------------------------
MAE using size to predict :4.558531024817282<br>
MAE using month to born to predict :6.929419052240534<br>
MAE using month to born and size to predict :4.61088154258837
---------------------------------------------------------------------------
MAPE using size to predict :0.09270083592592529<br>
MAPE using month to born to predict :0.16089084049913616<br>
MAPE using month to born and size to predict :0.09374353731896855
